esp/esdllib/esdl_script.cpp

```diff
@@ -1562,6 +1667,64 @@ class CEsdlTransformOperationForEach : public CEsdlTransformOperationWithChildre
     }
 };
 
+class CEsdlTransformOperationSynchronize : public CEsdlTransformOperationWithChildren
+{
+    Owned<ICompiledXpath> m_maxAtOnce;
+public:
+    CEsdlTransformOperationSynchronize(IXmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationWithChildren(xpp, stag, prefix, false, nullptr)
+    {
+        const char *maxAtOnce = stag.getValue("max-at-once");
+        if (!isEmptyString(maxAtOnce))
+            m_maxAtOnce.setown(compileXpath(maxAtOnce));
+    }
+
+    virtual ~CEsdlTransformOperationSynchronize(){}
+
+    virtual bool exec(CriticalSection *externalCrit, IInterface *preparedForAsync, IEsdlScriptContext * scriptContext, IXpathContext * targetContext, IXpathContext * sourceContext) override
+    {
+        if (!m_children.length())
+            return false;
+
+        IPointerArray preps;
+        ForEachItemIn(i, m_children)
+            preps.append(m_children.item(i).prepareForAsync(scriptContext, targetContext, sourceContext));
+
+        class casyncfor: public CAsyncFor
+        {
+            CriticalSection synchronizeCrit;
+            IPointerArray &preps;
+            IArrayOf<IEsdlTransformOperation> &children;
+            IEsdlScriptContext *scriptContext = nullptr;
+            IXpathContext *targetContext = nullptr;
+            IXpathContext *sourceContext = nullptr;
+
+        public:
+            casyncfor(IPointerArray &_preps, IArrayOf<IEsdlTransformOperation> &_children, IEsdlScriptContext *_scriptContext, IXpathContext *_targetContext, IXpathContext *_sourceContext)
+                : preps(_preps), children(_children), scriptContext(_scriptContext), targetContext(_targetContext), sourceContext(_sourceContext)
+            {
+            }
+            void Do(unsigned i)
+            {
+                children.item(i).exec(&synchronizeCrit, preps.item(i), scriptContext, targetContext, sourceContext);
+            }
+        } afor(preps, m_children, scriptContext, targetContext, sourceContext);
+
+        unsigned maxAtOnce = m_maxAtOnce.get() ? (unsigned) sourceContext->evaluateAsNumber(m_maxAtOnce) : 5;
```

Comment:
Why 5 instead of the original 10? I'm not suggesting 10 is better than 5; just wondering what prompted the change.

Reply:
Once it was configurable I started thinking about how many threads a single transaction should take for parallel processing by default.  This keeps it simple.  When used properly it would probably be rare to exceed this anyway, but users can increase the value if they want.